bba/bba.go

```diff
@@ -80,44 +87,117 @@ func (b *BBA) Result() bool {
 }
 
 func (b *BBA) Close() {
-	if first := atomic.CompareAndSwapInt32(&b.stop, int32(0), int32(1)); !first {
+	if first := atomic.CompareAndSwapInt32(&b.stopFlag, int32(0), int32(1)); !first {
 		return
 	}
 	b.closeChan <- struct{}{}
 }
 
-func (b *BBA) muxRequest(req request) error {
-	switch req.data.(type) {
-	case *bvalRequest:
-		return b.handleBvalRequest(req)
-	case *auxRequest:
-		return b.handleAuxRequest(req)
+func (b *BBA) muxMessage(id cleisthenes.ConnId, msg *pb.Message_Bba) error {
+	// TODO: request epoch check
+	switch msg.Bba.Type {
+	case pb.BBA_BVAL:
+		bval := &bvalRequest{}
+		if err := json.Unmarshal(msg.Bba.Payload, bval); err != nil {
+			return err
+		}
+		return b.handleBvalRequest(id, bval)
+	case pb.BBA_AUX:
+		aux := &auxRequest{}
+		if err := json.Unmarshal(msg.Bba.Payload, aux); err != nil {
+			return err
+		}
+		return b.handleAuxRequest(id, aux)
 	default:
-		return errors.New(fmt.Sprintf("unexpected request type"))
+		return ErrUndefinedRequestType
+	}
+	return ErrUndefinedRequestType
+}
+
+func (b *BBA) handleBvalRequest(id cleisthenes.ConnId, bval *bvalRequest) error {
+	if err := b.saveBvalIfNotExist(id, bval); err != nil {
+		return err
+	}
+	count := b.countBvalWithValue(bval.Value)
+	if count == 2*b.f+1 {
+		b.binValueSet.union(bval.Value)
+		b.binValueChan <- struct{}{}
+		return nil
+	}
+	if count == b.f+1 && !b.broadcastedBvalSet.exist(bval.Value) {
+		b.broadcastedBvalSet.union(bval.Value)
+		return b.broadcast(pb.BBA_BVAL, bval)
 	}
 	return nil
 }
 
-func (b *BBA) handleBvalRequest(msg request) error {
+func (b *BBA) handleAuxRequest(id cleisthenes.ConnId, req *auxRequest) error {
 	return nil
 }
 
-func (b *BBA) handleAuxRequest(msg request) error {
+func (b *BBA) broadcast(typ pb.BBAType, req cleisthenes.Request) error {
+	pl, err := json.Marshal(req)
```

Comment:
`pl` is too short. Until i see Payload, i didn't understand. I think just `payload` is enough.